src/Common/src/Internal/Text/Utf8String.cs

```diff
@@ -35,5 +38,20 @@ public override string ToString()
         {
             return Encoding.UTF8.GetString(_value);
         }
+
+        public override bool Equals(object obj)
+        {
+            return (obj is Utf8String) && Equals((Utf8String)obj);
+        }
+
+        public override int GetHashCode()
+        {
+            return ((IStructuralEquatable)_value).GetHashCode(EqualityComparer<byte>.Default);
```

Comment:
This is slow and it is not a very good hash function (it is only hashing last 8 elements). Could you please make this something better? E.g. something like:

```
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static int _rotl(int value, int shift)
{
    // This is expected to be optimized into a single rotl instruction
    return (int)(((uint)value << shift) | ((uint)value >> (32 - shift)));
}

public bool GetHashCode()
{
    unsafe
    {
        int length = _value.Length;
        int hash = length;
        fixed (byte * ap = _value)
        {
            byte* a = ap;

            while (length >= 4)
            {
                hash = (hash + _rotl(hash, 5)) ^ *(int*)a;
                a += 4; b += 4;
            }
            if (length >= 2)
            {
                hash = (hash + _rotl(hash, 5)) ^ *(short*)a;
                a += 2;
            }
            if (length > 0)
            {
                hash = (hash + _rotl(hash, 5)) ^ *a;
            }
            hash += _rotl(hash, 7);
            hash += _rotl(hash, 15);
            return hash;
        }
    }
}

public bool Equals(Utf8String other)
{
    int length = _value.Length;
    if (length != other.Length)
        return false;

    if (_value == other._value)
        return true;

    unsafe
    {
        fixed (byte * ap = _value) fixed (byte * bp = other._value)
        {
            byte* a = ap;
            byte* b = bp;

            while (length >= 4)
            {
                if (*(int*)a != *(int*)b) return false;
                a += 4; b += 4;
            }
            if (length >= 2)
            {
                if (*(short)a != *(short*)b) return false;
                a += 2; b += 2;
            }
            if (length > 0)
            {
                if (*a != *b) return false;
            }
            return true;
        }
    }
}
```